callisto/delivery/hashers.py

```diff
@@ -0,0 +1,209 @@
+    hashers = []
+    for hasher_path in settings.KEY_HASHERS:
+        hasher_cls = import_string(hasher_path)
+        hasher = hasher_cls()
+        if not getattr(hasher, 'algorithm'):
+            raise ImproperlyConfigured("hasher doesn't specify an algorithm name: {}".format(hasher_path))
+        hashers.append(hasher)
+    return hashers
+
+
+def get_hashers_by_algorithm():
+    hashers = get_hashers()
+    return {hasher.algorithm: hasher for hasher in hashers}
+
+
+def get_hasher(algorithm='default'):
+    if algorithm == 'default':
+        return get_hashers()[0]
+    else:
+        hashers = get_hashers_by_algorithm()
+        try:
+            return hashers[algorithm]
+        except KeyError:
+            raise ValueError("Unknown key hashing algorithm {0}."
+                             "Did you specify it in the KEY_HASHERS setting?".format(algorithm))
+
+
+def identify_hasher(encoded):
+    """
+    Returns a hasher based on either a fully encoded key or just the encode
+    prefix. If the encoded prefix is empty, assume the previous default hasher.
+    """
+    # assume all previous entries before this scheme is implemented use PBKDF2 + SHA256
+    if not encoded:
+        algorithm = 'pbkdf2_sha256'
+    else:
+        algorithm = encoded.split('$', 1)[0]
+    return get_hasher(algorithm)
+
+
+def make_key(encode_prefix, key, salt):
+    iterations = None
+    hasher = identify_hasher(encode_prefix)
+
+    if not encode_prefix:
+        assert salt is not None
+        iterations = settings.ORIGINAL_KEY_ITERATIONS
+    else:
+        salt = encode_prefix.rsplit('$', 1)[1]
+
+    if encode_prefix and hasher.algorithm == 'pbkdf2_sha256':
+        iterations = encode_prefix.split('$')[1]
+
+    encoded = hasher.encode(key, salt, iterations=iterations)
+    if hasher.algorithm == 'pbkdf2_sha256' and hasher.must_update(encode_prefix):
+        hasher.harden_runtime(key, encoded)
+
+    prefix, key = hasher.split_encoded(encoded)
+    return prefix, key
+
+
+class PBKDF2KeyHasher(PBKDF2PasswordHasher):
+    """
+    Key stretching using Django's PBKDF2 + SHA256 implementation.
+
+    Iterations may be changed safely in settings.
+    """
+    iterations = settings.KEY_ITERATIONS
+
+    def must_update(self, encode_prefix):
+        if not encode_prefix:
+            iterations = settings.ORIGINAL_KEY_ITERATIONS
+        else:
+            algorithm, iterations, salt = encode_prefix.split('$', 2)
+        return int(iterations) != self.iterations
+
+    def split_encoded(self, encoded):
+        """
+        Splits the encoded string into a separate prefix and stretched key.
+
+        Returns a prefix and a stretched key.
+        """
+        prefix, b64stretched = encoded.rsplit('$', 1)
+        stretched_key = base64.b64decode(b64stretched)
+        return prefix, force_bytes(stretched_key)
+
+
+# Portions of the below implementation are copyright the Django Software Foundation and individual contributors, and
```

Comment:
since it's "portions", let's move this to the top of the file to cover `get_hashers` etc